core/src/main/java/io/grpc/ResolvedServerInfo.java

```diff
@@ -73,4 +79,74 @@ public Attributes getAttributes() {
   public String toString() {
     return "[address=" + address + ", attrs=" + attributes + "]";
   }
+
+  /**
+   * Returns true if the given object is also a {@link ResolvedServerInfo} with an equal address
+   * and equal attribute values.
+   *
+   * <p>Two objects have equal attribute values if their {@linkplain #getAttributes() attributes}
+   * have the same set of {@linkplain Attributes#keys() keys} and the values associated with each
+   * key are equal.
+   *
+   * <p>Note that if a resolver includes mutable values in the attributes, it is possible for two
+   * objects to be considered equal at one point in time and not equal at another (due to concurrent
+   * mutation of attribute values).
+   *
+   * @param o an object
+   * @return true if the given object is a {@link ResolvedServerInfo} with an equal address and
+   * equal attributes
+   */
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) return true;
+    if (o == null || getClass() != o.getClass()) return false;
+    ResolvedServerInfo that = (ResolvedServerInfo) o;
+    return Objects.equal(address, that.address) &&
+        areAttributesEqual(attributes, that.attributes);
+  }
+
+  /**
+   * Returns a hash code for the server info. The hash code is computed like so:<pre>{@code
+   * address.hashCode() * 31 + hash(attributes)
+   * }</pre>
+   *
+   * <p>The hash of the attributes is computed as if the attributes were a {@link Map#hashCode()
+   * Map} whose entries consist of all of the attributes' {@linkplain Attributes#keys() keys} and
+   * their associated {@linkplain Attributes#get(Key) values}.
+   *
+   * <p>Note that if a resolver includes mutable values in the attributes, this object's hash code
+   * could change over time. So care must be used when putting these objects into a set or using
+   * them as keys for a map.
+   *
+   * @return a hash code for the server info, computed as described above
+   */
+  @Override
+  public int hashCode() {
+    return address.hashCode() * 31 + attributesHashCode(attributes);
+  }
+
+  private static boolean areAttributesEqual(Attributes a1, Attributes a2) {
```

Comment:
I see, you are avoiding defining hashCode on `Attributes`. That makes it even lower risk, but still usable for you. Sounds fine.